[TOPIC-GPIO] Update the PCAL9535A driver and rename it to PCA95XX

dcpleung opened the pull request:
This updates the PCAL9535A driver to the new GPIO API.

Also, rename the driver to PCA95XX since the driver can support that series of I2C-based GPIO expanders. Kconfig can be used to distinguish between variants.

Tested on the PCA9555 chip on the `mec15xxevb_assy6853` board.

dcpleung committed:
drivers: gpio_pcal9535a: update to use new GPIO API

Update driver code and board files to use new GPIO configuration flags
such as GPIO_ACTIVE_LOW. Also add implementation of new port_* driver
API as well as gpio_pin_interrupt_configure function.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: gpio: rename PCAL9535A to PCA95XX

PCA95XX is a series of compatible I2C-based GPIO expanders,
with common registers on input/output, polarity and configuration.
This renames the original PCAL9535A driver to PCA95XX to indicate
that it can support this series. Additional features on variants
are guarded by kconfigs.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Relates to #18530

pabigot reviewed (commented):
IMO this really needs to be converted to use devicetree instead of Kconfig for all its features, and distinct compatibles used to identify variants, per #19904.

Looks pretty trivial: a standard i2c-device just like `semtech,sx1409b` but adding a `has-pud` boolean.

dcpleung committed:
boards: mec15xxevb_assy6853: add DTS for PCA9555

There is a GPIO expander connected to I2C0 on board.
This adds the DTS entry.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
dts: add binding for PCA95xx series I2C-based GPIO expanders

This adds the DTS binding for the PCA95xx series I2C-based GPIO
expanders.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
gpio: pca95xx: convert kconfig to DTS

This converts configuration of the driver to DTS.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Updated to do DTS instead of Kconfigs. However, I cannot have multiple items in the compatible line in the binding, so I simply use "nxp,pca95xx" instead of actual chip model.

pabigot commented on boards/arm/mec15xxevb_assy6853/dts_fixup.h:
```diff
@@ -0,0 +1,26 @@
+/*
```
This file is not necessary if you use the devicetree names in the driver.

dcpleung replied:
Removed.

pabigot commented on drivers/gpio/gpio_pca95xx.c:
```diff
@@ -0,0 +1,679 @@
+
+/* Initialization for PCA95XX_0 */
+#ifdef CONFIG_GPIO_PCA95XX_0
+static const struct gpio_pca95xx_config gpio_pca95xx_0_cfg = {
```
This should be:
```
	.i2c_master_dev_name = DT_INST_0_NXP_PCA95XX_BUS_NAME,
￼	.i2c_slave_addr = DT_INST_0_NXP_PCA95XX_BASE_ADDRESS,
```

pabigot commented on drivers/gpio/gpio_pca95xx.c:
```diff
@@ -0,0 +1,679 @@
+static const struct gpio_pca95xx_config gpio_pca95xx_0_cfg = {
+	.i2c_master_dev_name = DT_GPIO_PCA95XX_0_I2C_MASTER_DEV_NAME,
+	.i2c_slave_addr = DT_GPIO_PCA95XX_0_I2C_ADDR,
+	.capabilities =
```
This can be:
```
    (DT_INST_0_NXP_PCA95XX_HAS_PUD ?  PCA_HAS_PUD : 0) |
```

pabigot commented on drivers/gpio/gpio_pca95xx.c:
```diff
@@ -0,0 +1,679 @@
+};
+
+/* This has to init after I2C master */
+DEVICE_AND_API_INIT(gpio_pca95xx_0, DT_GPIO_PCA95XX_0_DEV_NAME,
```
This should be `DT_INST_0_NXP_PCA95XX_LABEL`

pabigot reviewed (changes requested):
I'm still focusing on devicetree support here, since I don't have hardware to test this.  But can you confirm that your testing showed that both the 1pin and basic GPIO tests pass, as well as blinky and button (if appropriate)?

The native devicetree names should be used rather than the aliases that were necessary in the Kconfig era.  I've noted substitutes for the first instance, but they should be applied to all instances.

If there's actually a need for so many instances you might look a `drivers/counter/counter_nrfx_rtc.c` for an example of a macro that generates all the instance-specific boilerplate just given the instance number.

pabigot commented:
> I cannot have multiple items in the compatible line in the binding

True; that'd require something like #20289.

You should be able to do:

    compatible = "nxp,pca9555", "nxp,pca95xx";

in the devicetree source, though you'll get warnings from checkpatch until #19904 is settled.

dcpleung commented:
Updated the driver to use 'DT_INST_*`. As for testing, I have only access to the mec15xx board with PCA9555. This variant does not support pull-up/pull-down so almost all test cases in the 1 pin test are skipped (except the LED toggle which works on the board). The driver does not support any interrupts so the button test would not work. The GPIO basic test, on the other hand, passed with the pull-up/pull-down and interrupt tests skipped.

pabigot commented on drivers/gpio/gpio_pca95xx.c:
```diff
@@ -0,0 +1,671 @@
+#define PCA_HAS_PUD			BIT(0)
+
+/**
+ * @brief Check to see if a I2C master is identified for communication.
```
This really shouldn't be necessary.  The only way to get a device pointer is for initialization to have succeeded, in which case `i2c_master` would have been found.

mnkp replied:
Seems like the idea behind the function is to make sure that none of the public API functions will attempt to communicate over i2c bus if the initialization of the i2c master device has failed. But the name of the function is rather confusing.

dcpleung replied:
The function is removed.

pabigot commented on drivers/gpio/gpio_pca95xx.c:
```diff
@@ -0,0 +1,671 @@
+	int ret;
+
+	/* For each pin, 0 == output, 1 == input */
+	switch (access_op) {
```
This switch is unnecessary since the function can only be reached when `access_op` is `GPIO_ACCESS_BY_PIN` or an invalid value: `GPIO_ACCESS_BY_PORT` is no longer passed into the driver API, and it's not passed internally in configuration operations.

dcpleung replied:
I keep these here because `gpio_config()` is still a valid public API, where `GPIO_ACCESS_BY_PORT` is still a valid `access_op`. These switch statements can be collapsed into simply `GPIO_ACCESS_BY_PIN` once the old APIs are removed.

mnkp replied:
We would like to remove the `access_op` very soon. It would help us if the code was written in a way that makes it easy. Already today no API function will be called with `GPIO_ACCESS_BY_PORT`. It would be enough if at the beginning of `gpio_pca95xx_config` function you add the following code. Then all the dependencies can be cleaned up already.
```
	if (access_op != GPIO_ACCESS_BY_PIN) {
		return -ENOTSUP;
	}
```
I understand it's more work but it will make reviewing the code easier.

dcpleung replied:
No problem. I have uploaded a new version.

pabigot commented on drivers/gpio/gpio_pca95xx.c:
```diff
@@ -0,0 +1,671 @@
+
+	/* Setup pin pull up or pull down */
+	port = &drv_data->reg_cache.pud_sel;
+	switch (access_op) {
```
As with previous function: validate against and implement `GPIO_ACCESS_BY_PIN` only.

Also applies to switch statement later in this function.

dcpleung replied:
(as above about `gpio_config()` is still a valid public API.)

pabigot reviewed (commented):
The code generally looks OK for the basic operations, but you should look at the sx1509b driver implementation for examples of:
* locking to prevent corrupting the configuration or output state if two threads try to manipulate the pins or configuration at the same time;
* checks to block operations from being initiated from within an ISR.

We don't document requirements for atomicity and interrupt-related behavior but for external controllers they can be a serious problem.

Also port-based configuration code can be dropped.

mnkp commented on drivers/gpio/gpio_pca95xx.c:
```diff
@@ -0,0 +1,671 @@
+	if (ret) {
+		LOG_ERR("PCA95XX[0x%X]: error writing to register 0x%X "
+				"(%d)", i2c_addr, reg, ret);
+	}
```
Performing two independent i2c bus write transactions is a bit expensive. It would be better to use `i2c_burst_write`. According to the datasheet it's supported by the chip. We only need to be careful that `buf` payload is sent correctly on big-endian and little-endian MCUs.

mnkp commented on drivers/gpio/gpio_pca95xx.c:
```diff
@@ -0,0 +1,671 @@
+#define REG_INT_MASK_PORT1		0x4B
+#define REG_INT_STATUS_PORT0		0x4C
+#define REG_INT_STATUS_PORT1		0x4D
+#define REG_OUTPUT_PORT_CONF		0x4F
```
Since we already have `gpio_pca95xx.h` header it would be better to put these defines there.

mnkp commented on drivers/gpio/gpio_pca95xx.c:
```diff
@@ -0,0 +1,671 @@
+
+	new_value = (port->all & ~mask) | (mask & value);
+
+	return gpio_pca95xx_write(dev, GPIO_ACCESS_BY_PORT, 0, new_value);
```
The `GPIO_ACCESS_BY_PORT` will be soon removed from the API and - as mentioned by @pabigot - it's already unused. All the new `gpio_pca95xx_port_*` functions should use `write_port_regs`, `read_port_regs` directly.

dcpleung replied:
Does this mean that the current `.read()` and `.write()` APIs are going away, and that read/write by pin will no longer be supported?

mnkp replied:
Yes, the `.read()` and `.write()` functions are going away. The API calls will be deprecated in this release and removed later. We want to remove their implementations already. The API will call new functions instead in a way that should be transparent to the user.

dcpleung replied:
In that case, I will duplicate the current read/write functions without the port access. So when it is time to remove the API, those old functions can be trivially removed.

mnkp replied:
I think it will be cleaner if you call here `write_port_regs` directly. Only the mutex logic needs to be moved inside `write_port_regs` as well.

mnkp replied:
We need also to update the cached value of the output register. Doing it in a way that's thread safe is a bit tricky. You may want to check `drivers/gpio/gpio_sx1509b.c` driver that has already been updated.

dcpleung replied:
I have updated the code to do this. So when GPIO_ACCESS_BY_PORT is remove, those functions can be removed trivially.

mnkp commented on drivers/gpio/gpio_pca95xx.c:
```diff
@@ -0,0 +1,671 @@
+{
+	struct gpio_pca95xx_drv_data * const drv_data =
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	union gpio_pca95xx_port_data *port = &drv_data->reg_cache.output;
```
Let's rename `port` to `reg_out`.

mnkp commented on dts/bindings/gpio/nxp,pca95xx.yaml:
```diff
@@ -0,0 +1,27 @@
+    label:
+      required: true
+
+    "has-pud":
```
Property name should be without a quotation mark "

dcpleung committed:
drivers: gpio_pca95xx: use I2C burst write

Use I2C burst write to write 2 bytes to each pair of registers
instead of 2 separate transactions of writing 1 byte.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: gpio_pca95xx: roll header file into source file

The structures defined in the header file are only used by
the driver source file, and should not be used by others.
So roll the header file into the source file so it won't
get #include.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: gpio_pca95xx: guard read/write with semaphore

Since the GPIO expander is a I2C device, any read/write to
the registers has high latency. Therefore, semaphore is
introduced to prevent multiple threads to manipulate
the GPIOs at the same time.

Also make sure that we are not doing I2C transactions
within ISRs.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

mnkp commented on drivers/gpio/gpio_pca95xx.c:
```diff
@@ -0,0 +1,695 @@
+	    POST_KERNEL, CONFIG_GPIO_PCA95XX_INIT_PRIORITY,
+	    &gpio_pca95xx_drv_api_funcs);
+
+#endif /* CONFIG_GPIO_PCA95XX_0 */
```
```
#endif /* DT_INST_0_NXP_PCA95XX */
```

dcpleung replied:
Forgot about it... uploaded a new version.

dcpleung committed:
drivers: gpio_pca95xx: no longer support config by port

Remove support for GPIO_ACCESS_BY_PORT in the config function
as configuration by port is going away.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

mnkp reviewed (commented):
@dcpleung thanks for the update. It would be really helpful if you removed redundant usage of `access_op`.

mnkp commented on drivers/gpio/gpio_pca95xx.c:
```diff
@@ -0,0 +1,675 @@
+		(struct gpio_pca95xx_drv_data * const)dev->driver_data;
+	union gpio_pca95xx_port_data *reg_out = &drv_data->reg_cache.output;
+	u32_t new_value;
+
```
We need to add
```
	if (k_is_in_isr()) {
		return -EWOULDBLOCK;
	}
```
to all the new API functions.

mnkp commented on drivers/gpio/gpio_pca95xx.c:
```diff
@@ -0,0 +1,675 @@
+	union gpio_pca95xx_port_data buf;
+	int ret;
+
+	ret = read_port_regs(dev, REG_INPUT_PORT0, &buf);
```
Reading over i2c bus needs to be guarded by mutex, like in the case of write.

mnkp commented on drivers/gpio/gpio_pca95xx.c:
```diff
@@ -0,0 +1,675 @@
+		"0x%X", i2c_addr, reg, buf->byte[0], (reg + 1),
+		buf->byte[1]);
+
+	ret = i2c_burst_write(i2c_master, i2c_addr, reg, buf->byte, 2);
```
We need to be careful here since on a big endian MCU the value in `buf->byte` will have LSB located in `byte[1]`. Check implementation of `i2c_reg_write_word_be` function in `gpio_sx1509b.c` driver. The difference is that here the value needs to be little endian (if I read PCAL9535A datasheet correctly). The same holds true for the `read_port_regs` function.

mnkp commented on drivers/gpio/gpio_pca95xx.c:
```diff
@@ -0,0 +1,675 @@
+	union gpio_pca95xx_port_data *reg_out = &drv_data->reg_cache.output;
+	u32_t new_value;
+
+	new_value = reg_out->all | mask;
```
If the thread gets preempted the moment we read out `reg_out` value its value can be changed by another thread. Subsequently, if this thread updates the value of `reg_out`, which happens inside `write_out_regs_locked`, the value placed there by another thread will be overwritten. We have a race condition. The full read, modify, write operation needs to be protected by the mutex. A good example of how this can be implemented can be found in `gpio_sx1509b.c` driver.

dcpleung committed:
drivers: gpio_pca95xx: endianness awareness

The register pair for each port in the GPIO expander are
port 0 first then port 1. This would not work for big
endian systems with the u16_t port value. So need to
swap the byte ordering on such system.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: gpio_pca95xx: update cache after successful write

Only update the internal register cache after successful write,
or else it would get out of sync with hardware.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
drivers: gpio_pca95xx: factor device declaration into macros

This factors the common bits of device declaration into macros
so it would be easier to add new instances.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

mnkp reviewed (commented):
LGTM, with a few minor comments.
